{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-08-10T10:03:00.595966Z",
     "start_time": "2018-08-10T10:03:00.587806Z"
    }
   },
   "source": [
    "# Meth_Comp API usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-07T14:13:14.870400Z",
     "start_time": "2020-04-07T14:13:14.032380Z"
    },
    "init_cell": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Import main module \n",
    "from pycoMeth.Comp_Report import Comp_Report\n",
    "\n",
    "# optionally inport jupyter helper functions\n",
    "from pycoMeth.common import head, jhelp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-07T14:13:14.882221Z",
     "start_time": "2020-04-07T14:13:14.871885Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**Comp_Report** (methcomp_fn, gff3_fn, outdir, n_top, max_tss_distance, pvalue_threshold, min_diff_llr, verbose, quiet, progress, kwargs)\n",
       "\n",
       "\n",
       "\n",
       "---\n",
       "\n",
       "* **methcomp_fn** (required) [str]\n",
       "\n",
       "Input tsv file generated by Meth_comp (can be gzipped). At the moment only data binned by intervals with Interval_Aggregate are supported.\n",
       "\n",
       "* **gff3_fn** (required) [str]\n",
       "\n",
       "Path to an **ensembl GFF3** file containing genomic annotations. Only the transcripts details are extracted.\n",
       "\n",
       "* **outdir** (default: \"\") [str]\n",
       "\n",
       "Directory where to output HTML reports, By default current directory\n",
       "\n",
       "* **n_top** (default: 50) [int]\n",
       "\n",
       "Number of top interval candidates for which to generate an interval report. If there are not enough significant candidates this is automatically scaled down.\n",
       "\n",
       "* **max_tss_distance** (default: 100000) [int]\n",
       "\n",
       "Maximal distance to transcription stat site to find transcripts close to interval candidates\n",
       "\n",
       "* **pvalue_threshold** (default: 0.01) [float]\n",
       "\n",
       "pValue cutoff for top interval candidates\n",
       "\n",
       "* **min_diff_llr** (default: 1) [float]\n",
       "\n",
       "Minimal llr boundary for negative and positive median llr. 1 is recommanded for vizualization purposes.\n",
       "\n",
       "* **verbose** (default: False) [bool]\n",
       "\n",
       "* **quiet** (default: False) [bool]\n",
       "\n",
       "* **progress** (default: False) [bool]\n",
       "\n",
       "* **kwargs**\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(Comp_Report)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Example with a single significant result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-07T14:13:21.209471Z",
     "start_time": "2020-04-07T14:13:19.469317Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m## Checking options and input files ##\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: Options summary\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tPackage name: pycoMeth\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tPackage version: 0.4.0\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tTimestamp: 2020-04-07 15:13:19.471381\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tkwargs\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tprogress: False\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tquiet: False\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tverbose: True\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tmin_diff_llr: 1\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tpvalue_threshold: 0.05\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tmax_tss_distance: 100000\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tn_top: 50\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \toutdir: yeast_html\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tgff3_fn: ./data/yeast.gff3\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: \tmethcomp_fn: ./data/Yeast_CGI_meth_comp.tsv.gz\u001b[0m\n",
      "\u001b[01;34m## Loading and preparing data ##\u001b[0m\n",
      "\u001b[32m\tLoading Methcomp data from TSV file\u001b[0m\n",
      "\u001b[32m\tLoading transcript info from GFF file\u001b[0m\n",
      "\u001b[32m\tNumber of significant intervals found (adjusted pvalue<0.05): 1\u001b[0m\n",
      "\u001b[01;31mERROR: Low number of significant sites. The summary report will likely contains errors\u001b[0m\n",
      "\u001b[01;31mERROR: Number of significant intervals lower than number of top candidates to plot\u001b[0m\n",
      "\u001b[32m\tGenerating file names for top candidates reports\u001b[0m\n",
      "\u001b[32m\tComputing source md5\u001b[0m\n",
      "\u001b[01;34m## Parsing methcomp data ##\u001b[0m\n",
      "\u001b[32m\tIterating over significant intervals and generating top candidates reports\u001b[0m\n",
      "\u001b[37m\t[DEBUG]: Ploting top candidates: V-65,380-65,612\u001b[0m\n",
      "\u001b[32m\tGenerating summary report\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "Comp_Report (\n",
    "    methcomp_fn = \"./data/Yeast_CGI_meth_comp.tsv.gz\",\n",
    "    gff3_fn = \"./data/yeast.gff3\",\n",
    "    outdir = \"yeast_html\",\n",
    "    pvalue_threshold = 0.05,\n",
    "    verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Usage with large dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-07T14:13:56.452332Z",
     "start_time": "2020-04-07T14:13:23.787913Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m## Checking options and input files ##\u001b[0m\n",
      "\u001b[01;34m## Loading and preparing data ##\u001b[0m\n",
      "\u001b[32m\tLoading Methcomp data from TSV file\u001b[0m\n",
      "\u001b[32m\tLoading transcript info from GFF file\u001b[0m\n",
      "\u001b[01;31mERROR: Not all the chromosomes found in the data file are present in the GFF3 file. This will lead to missing transcript ids\u001b[0m\n",
      "\u001b[32m\tNumber of significant intervals found (adjusted pvalue<0.01): 3532\u001b[0m\n",
      "\u001b[32m\tGenerating file names for top candidates reports\u001b[0m\n",
      "\u001b[32m\tComputing source md5\u001b[0m\n",
      "\u001b[01;34m## Parsing methcomp data ##\u001b[0m\n",
      "\u001b[32m\tIterating over significant intervals and generating top candidates reports\u001b[0m\n",
      "\tProgress: 100%|██████████| 3.53k/3.53k [00:29<00:00, 122 intervals/s] \n",
      "\u001b[32m\tGenerating summary report\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "Comp_Report (\n",
    "    methcomp_fn = \"./data/Medaka_CGI_meth_comp.tsv.gz\",\n",
    "    gff3_fn = \"./data/medaka.gff3\",\n",
    "    outdir = \"medaka_html\",\n",
    "    n_top=25,\n",
    "    progress=True)"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "hide_input": false,
  "kernelspec": {
   "display_name": "pycoMeth",
   "language": "python",
   "name": "pycometh"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
