{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-08-10T10:03:00.595966Z",
     "start_time": "2018-08-10T10:03:00.587806Z"
    }
   },
   "source": [
    "# Interval_Aggregate API usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-07T13:14:32.446501Z",
     "start_time": "2020-04-07T13:14:32.442281Z"
    },
    "init_cell": true
   },
   "outputs": [],
   "source": [
    "# Import main module \n",
    "from pycoMeth.Interval_Aggregate import Interval_Aggregate\n",
    "\n",
    "# optionally inport jupyter helper functions\n",
    "from pycoMeth.common import head, jhelp, stdout_print"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-07T13:14:33.822722Z",
     "start_time": "2020-04-07T13:14:33.799230Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**Interval_Aggregate** (cpg_aggregate_fn, ref_fasta_fn, interval_bed_fn, output_bed_fn, output_tsv_fn, interval_size, min_cpg_per_interval, sample_id, min_llr, verbose, quiet, progress, kwargs)\n",
       "\n",
       "Bin the output of `pycoMeth CpG_Aggregate` in genomic intervals, using either an annotation file containing intervals or a sliding window.\n",
       "\n",
       "---\n",
       "\n",
       "* **cpg_aggregate_fn** (required) [str]\n",
       "\n",
       "Output tsv file generated by CpG_Aggregate (can be gzipped)\n",
       "\n",
       "* **ref_fasta_fn** (required) [str]\n",
       "\n",
       "Reference file used for alignment in Fasta format (ideally already indexed with samtools faidx)\n",
       "\n",
       "* **interval_bed_fn** (default: None) [str]\n",
       "\n",
       "SORTED bed file containing **non-overlapping** intervals to bin CpG data into (Optional) (can be gzipped)\n",
       "\n",
       "* **output_bed_fn** (default: None) [str]\n",
       "\n",
       "Path to write a summary result file in BED format (At least 1 output file is required) (can be gzipped)\n",
       "\n",
       "* **output_tsv_fn** (default: None) [str]\n",
       "\n",
       "Path to write a more extensive result report in TSV format (At least 1 output file is required) (can be gzipped)\n",
       "\n",
       "* **interval_size** (default: 1000) [int]\n",
       "\n",
       "Size of the sliding window in which to aggregate CpG sites data from if no BED file is provided\n",
       "\n",
       "* **min_cpg_per_interval** (default: 5) [int]\n",
       "\n",
       "Minimal number of CpG sites per interval.\n",
       "\n",
       "* **sample_id** (default: \"\") [str]\n",
       "\n",
       "Sample ID to be used for the BED track header\n",
       "\n",
       "* **min_llr** (default: 2) [float]\n",
       "\n",
       "Minimal log likelyhood ratio to consider a site significantly methylated or unmethylated in output BED file\n",
       "\n",
       "* **verbose** (default: False) [bool]\n",
       "\n",
       "* **quiet** (default: False) [bool]\n",
       "\n",
       "* **progress** (default: False) [bool]\n",
       "\n",
       "* **kwargs**\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp (Interval_Aggregate)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Default usage with sliding windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-07T13:14:39.559357Z",
     "start_time": "2020-04-07T13:14:37.009646Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m## Checking options and input files ##\u001b[0m\n",
      "\u001b[01;34m## Parsing CpG_aggregate file ##\u001b[0m\n",
      "\tProgress: 100%|██████████| 5.82M/5.82M [00:02<00:00, 2.30M bytes/s]\n",
      "\u001b[32m\tResults summary\u001b[0m\n",
      "\u001b[32m\t\tLines parsed: 89,392\u001b[0m\n",
      "\u001b[32m\t\tTotal number of intervals: 24,319\u001b[0m\n",
      "\u001b[32m\tWritter summary\u001b[0m\n",
      "\u001b[32m\t\tEmpty intervals skipped: 14,390\u001b[0m\n",
      "\u001b[32m\t\tValid intervals written: 9,195\u001b[0m\n",
      "\u001b[32m\t\tLow CpG intervals skipped: 734\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chromosome start end  num_motifs median_llr llr_list                                           pos_list                                           \n",
      "I          500   1000 12         -3.35      [-1.14,-3.54,-7.24,-4.3,0.56,-0.65,-4.37,-3.78,...[557,587,628,665,834,868,890,936,955,967,988]      \n",
      "I          1000  1500 22         -3.65      [-2.48,-5.035,-4.16,-3.315,-3.295,-1.69,-9.885,...[1036,1095,1119,1136,1158,1178,1199,1217,1345,1...\n",
      "I          1500  2000 19         -3.4       [-5.71,-6.05,-0.925,-7.165,-3.975,0.56,-1.78,-1...[1523,1584,1630,1654,1707,1755,1784,1797,1814,1...\n",
      "I          2000  2500 15         -4.272     [-5.24,-3.07,-4.33,-19.055,-7.55,-1.255,-2.565,...[2003,2051,2084,2137,2302,2396,2421,2445,2462,2...\n",
      "I          2500  3000 19         -2.5       [-0.705,0.385,-6.685,-10.175,-4.27,-2.3,-2.5,-2...[2546,2563,2584,2634,2666,2680,2694,2729,2752,2...\n",
      "I          3000  3500 9          -1.4       [-2.34,-2.19,-1.4,-1.0,0.29,-1.3,-6.9,-1.22,-1.92] [3000,3024,3044,3056,3071,3148,3218,3367,3473]     \n",
      "I          3500  4000 8          -1.75      [-1.7,-0.53,-4.46,-2.2,-1.75,-8.47,0.53]           [3516,3610,3624,3674,3722,3823,3987]               \n",
      "I          4000  4500 10         -3.325     [-2.83,-3.96,-1.77,-13.895,-3.82,-1.73]            [4078,4094,4223,4276,4296,4399]                    \n",
      "I          4500  5000 12         -2.628     [-2.71,-0.76,-1.55,-2.56,-0.985,-5.02,-3.46,-1....[4534,4591,4654,4706,4787,4814,4847,4859,4917,4...\n",
      "\n",
      "track name=sample_1_Interval itemRgb=On\n",
      "I\t500\t1000\t.\t-3.35\t.\t500\t1000\t29,140,190\n",
      "I\t1000\t1500\t.\t-3.65\t.\t1000\t1500\t29,140,190\n",
      "I\t1500\t2000\t.\t-3.4\t.\t1500\t2000\t29,140,190\n",
      "I\t2000\t2500\t.\t-4.272\t.\t2000\t2500\t33,102,171\n",
      "I\t2500\t3000\t.\t-2.5\t.\t2500\t3000\t52,168,194\n",
      "I\t3000\t3500\t.\t-1.4\t.\t3000\t3500\t230,230,230\n",
      "I\t3500\t4000\t.\t-1.75\t.\t3500\t4000\t230,230,230\n",
      "I\t4000\t4500\t.\t-3.325\t.\t4000\t4500\t29,140,190\n",
      "I\t4500\t5000\t.\t-2.628\t.\t4500\t5000\t52,168,194\n",
      "\n"
     ]
    }
   ],
   "source": [
    "Interval_Aggregate (\n",
    "    cpg_aggregate_fn=\"./data/CpG_Aggregate_sample_1.tsv\",\n",
    "    ref_fasta_fn=\"./data/ref.fa\",\n",
    "    output_bed_fn=\"./results/Interval_Aggregate_sample_1.bed\",\n",
    "    output_tsv_fn=\"./results/Interval_Aggregate_sample_1.tsv\",\n",
    "    interval_size=500,\n",
    "    min_cpg_per_interval=3,\n",
    "    sample_id=\"sample_1\",\n",
    "    progress=True)\n",
    "\n",
    "head(\"./results/Interval_Aggregate_sample_1.tsv\")\n",
    "head(\"./results/Interval_Aggregate_sample_1.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Usage with a CpG Islands annotation Bed file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-07T13:14:40.665805Z",
     "start_time": "2020-04-07T13:14:39.561228Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m## Checking options and input files ##\u001b[0m\n",
      "\u001b[01;34m## Parsing CpG_aggregate file ##\u001b[0m\n",
      "\tProgress: 100%|█████████▉| 5.81M/5.82M [00:01<00:00, 5.35M bytes/s]\n",
      "\u001b[32m\tResults summary\u001b[0m\n",
      "\u001b[32m\t\tLines parsed: 89,235\u001b[0m\n",
      "\u001b[32m\t\tTotal number of intervals: 2,041\u001b[0m\n",
      "\u001b[32m\tWritter summary\u001b[0m\n",
      "\u001b[32m\t\tEmpty intervals skipped: 1,323\u001b[0m\n",
      "\u001b[32m\t\tValid intervals written: 718\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chromosome start end   num_motifs median_llr llr_list                                           pos_list                                           \n",
      "I          1804  2170  14         -3.67      [-3.67,-3.4,-5.53,-1.06,-1.79,-1.94,-6.22,-5.24...[1814,1829,1889,1925,1949,1961,1976,2003,2051,2...\n",
      "I          31835 32949 10         -5.65      [-2.925,-6.055,-1.785,-5.65,-6.83,-1.695,-12.32]   [31867,31889,31937,31960,32006,32031,32056]        \n",
      "I          33497 34371 19         -3.295     [-4.38,-3.32,-1.29,-3.27,-5.89,-8.96,-6.88,-2.2...[33947,33967,34001,34021,34049,34068,34099,3416...\n",
      "I          44730 44988 9          -3.2       [-2.37,-4.9,-1.63,-1.69,-8.09,-4.03]               [44748,44789,44808,44841,44877,44930]              \n",
      "I          47889 48187 13         -4.55      [-4.55,-9.41,-3.37,-4.66,-3.24,-4.66,-4.535]       [47897,48003,48036,48050,48084,48100,48115]        \n",
      "I          57175 57391 9          -4.76      [-7.96,-4.76,-0.33,-3.77,-1.03,-7.68,-6.66]        [57200,57255,57274,57292,57316,57335,57359]        \n",
      "I          59052 59257 13         -4.05      [-8.53,-1.23,-11.15,-3.07,-3.88,-4.22,-1.59,-5.79] [59071,59109,59142,59167,59187,59219,59232,59249]  \n",
      "I          60422 60656 18         -3.615     [-3.3,-3.14,-0.44,-6.65,-3.93,-2.27,-9.16,-1.73...[60427,60440,60467,60493,60519,60548,60561,6058...\n",
      "I          61246 61903 38         -3.423     [-3.68,-26.15,-6.45,-0.76,-3.86,-2.33,-4.945,-4...[61262,61353,61391,61409,61424,61443,61501,6154...\n",
      "\n",
      "track name=sample_1_Interval itemRgb=On\n",
      "I\t1804\t2170\t.\t-3.67\t.\t1804\t2170\t29,140,190\n",
      "I\t31835\t32949\t.\t-5.65\t.\t31835\t32949\t35,70,156\n",
      "I\t33497\t34371\t.\t-3.295\t.\t33497\t34371\t29,140,190\n",
      "I\t44730\t44988\t.\t-3.2\t.\t44730\t44988\t29,140,190\n",
      "I\t47889\t48187\t.\t-4.55\t.\t47889\t48187\t33,102,171\n",
      "I\t57175\t57391\t.\t-4.76\t.\t57175\t57391\t33,102,171\n",
      "I\t59052\t59257\t.\t-4.05\t.\t59052\t59257\t33,102,171\n",
      "I\t60422\t60656\t.\t-3.615\t.\t60422\t60656\t29,140,190\n",
      "I\t61246\t61903\t.\t-3.423\t.\t61246\t61903\t29,140,190\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ff = Interval_Aggregate (\n",
    "    cpg_aggregate_fn=\"./data/CpG_Aggregate_sample_1.tsv\",\n",
    "    ref_fasta_fn=\"./data/ref.fa\",\n",
    "    interval_bed_fn=\"./data/Yeast_CGI.bed\",\n",
    "    output_bed_fn=\"./results/CGI_Aggregate_sample_1.bed\",\n",
    "    output_tsv_fn=\"./results/CGI_Aggregate_sample_1.tsv\",\n",
    "    sample_id=\"sample_1\",\n",
    "    min_cpg_per_interval=1,\n",
    "    progress=True)\n",
    "\n",
    "head(\"./results/CGI_Aggregate_sample_1.tsv\")\n",
    "head(\"./results/CGI_Aggregate_sample_1.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example with multiple files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-07T13:14:52.344344Z",
     "start_time": "2020-04-07T13:14:43.316099Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##### SAMPLE 1 #####"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m## Checking options and input files ##\u001b[0m\n",
      "\u001b[01;34m## Parsing CpG_aggregate file ##\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##### SAMPLE 2 #####"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m## Checking options and input files ##\u001b[0m\n",
      "\u001b[01;34m## Parsing CpG_aggregate file ##\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##### SAMPLE 3 #####"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m## Checking options and input files ##\u001b[0m\n",
      "\u001b[01;34m## Parsing CpG_aggregate file ##\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##### SAMPLE 4 #####"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m## Checking options and input files ##\u001b[0m\n",
      "\u001b[01;34m## Parsing CpG_aggregate file ##\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "for i in range (1, 5):\n",
    "    stdout_print (f\"##### SAMPLE {i} #####\")\n",
    "    Interval_Aggregate (\n",
    "        cpg_aggregate_fn=f\"./data/CpG_Aggregate_sample_{i}.tsv\",\n",
    "        ref_fasta_fn=\"./data/ref.fa\",\n",
    "        output_bed_fn=f\"./results/Interval_Aggregate_sample_{i}.bed\",\n",
    "        output_tsv_fn=f\"./results/Interval_Aggregate_sample_{i}.tsv\",\n",
    "        sample_id=f\"sample_{i}\",\n",
    "        interval_size=500,\n",
    "        min_cpg_per_interval=3,\n",
    "        min_llr=1,\n",
    "        quiet=True)"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "hide_input": false,
  "kernelspec": {
   "display_name": "pycoMeth",
   "language": "python",
   "name": "pycometh"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
