{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-08-10T10:03:00.595966Z",
     "start_time": "2018-08-10T10:03:00.587806Z"
    }
   },
   "source": [
    "# Interval_Aggregate CLI usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Activate virtual environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-26T12:52:14.845101Z",
     "start_time": "2020-02-26T12:52:13.924808Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(pycoMeth) (pycoMeth) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Using virtualenvwrapper here but can also be done with Conda \n",
    "workon pycoMeth"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-26T12:52:16.961257Z",
     "start_time": "2020-02-26T12:52:15.669137Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: pycoMeth Interval_Aggregate [-h] -i CPG_AGGREGATE_FN -f REF_FASTA_FN\n",
      "                                   [-a INTERVAL_BED_FN] [-b OUTPUT_BED_FN]\n",
      "                                   [-t OUTPUT_TSV_FN] [-n INTERVAL_SIZE]\n",
      "                                   [-m MIN_CPG_PER_INTERVAL] [-s SAMPLE_ID]\n",
      "                                   [-l MIN_LLR] [-v] [-q] [-p]\n",
      "\n",
      "Bin the output of `pycoMeth CpG_Aggregate` in genomic intervals, using either\n",
      "an annotation file containing intervals or a sliding window.\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "\n",
      "Input/Output options:\n",
      "  -i CPG_AGGREGATE_FN, --cpg_aggregate_fn CPG_AGGREGATE_FN\n",
      "                        Output tsv file generated by CpG_Aggregate (can be\n",
      "                        gzipped) (required) [str]\n",
      "  -f REF_FASTA_FN, --ref_fasta_fn REF_FASTA_FN\n",
      "                        Reference file used for alignment in Fasta format\n",
      "                        (ideally already indexed with samtools faidx)\n",
      "                        (required) [str]\n",
      "  -a INTERVAL_BED_FN, --interval_bed_fn INTERVAL_BED_FN\n",
      "                        SORTED bed file containing **non-overlapping**\n",
      "                        intervals to bin CpG data into (Optional) (can be\n",
      "                        gzipped) (default: None) [str]\n",
      "  -b OUTPUT_BED_FN, --output_bed_fn OUTPUT_BED_FN\n",
      "                        Path to write a summary result file in BED format (At\n",
      "                        least 1 output file is required) (can be gzipped)\n",
      "                        (default: None) [str]\n",
      "  -t OUTPUT_TSV_FN, --output_tsv_fn OUTPUT_TSV_FN\n",
      "                        Path to write a more extensive result report in TSV\n",
      "                        format (At least 1 output file is required) (can be\n",
      "                        gzipped) (default: None) [str]\n",
      "\n",
      "Misc options:\n",
      "  -n INTERVAL_SIZE, --interval_size INTERVAL_SIZE\n",
      "                        Size of the sliding window in which to aggregate CpG\n",
      "                        sites data from if no BED file is provided (default:\n",
      "                        1000) [int]\n",
      "  -m MIN_CPG_PER_INTERVAL, --min_cpg_per_interval MIN_CPG_PER_INTERVAL\n",
      "                        Minimal number of CpG sites per interval. (default: 5)\n",
      "                        [int]\n",
      "  -s SAMPLE_ID, --sample_id SAMPLE_ID\n",
      "                        Sample ID to be used for the BED track header\n",
      "                        (default: None) [str]\n",
      "  -l MIN_LLR, --min_llr MIN_LLR\n",
      "                        Minimal log likelyhood ratio to consider a site\n",
      "                        significantly methylated or unmethylated in output BED\n",
      "                        file (default: 2) [float]\n",
      "\n",
      "Verbosity options:\n",
      "  -v, --verbose         Increase verbosity\n",
      "  -q, --quiet           Reduce verbosity\n",
      "  -p, --progress        Display a progress bar\n",
      "(pycoMeth) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoMeth Interval_Aggregate --help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Default usage with sliding windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-26T12:52:32.847818Z",
     "start_time": "2020-02-26T12:52:29.196983Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "## Checking options and input files ##\n",
      "## Parsing CpG_aggregate file ##\n",
      "\tResults summary\n",
      "\t\tLines parsed: 89,392\n",
      "\t\tTotal number of intervals: 24,319\n",
      "\tWritter summary\n",
      "\t\tEmpty intervals skipped: 14,390\n",
      "\t\tValid intervals written: 8,389\n",
      "\t\tLow CpG intervals skipped: 1,540\n",
      "(pycoMeth) (pycoMeth) track name=sample_1_Interval itemRgb=On\n",
      "I\t500\t1000\t.\t-3.350\t.\t500\t1000\t29,140,190\n",
      "I\t1000\t1500\t.\t-3.650\t.\t1000\t1500\t29,140,190\n",
      "I\t1500\t2000\t.\t-3.400\t.\t1500\t2000\t29,140,190\n",
      "I\t2000\t2500\t.\t-4.272\t.\t2000\t2500\t33,102,171\n",
      "I\t2500\t3000\t.\t-2.500\t.\t2500\t3000\t52,168,194\n",
      "I\t3000\t3500\t.\t-1.400\t.\t3000\t3500\t230,230,230\n",
      "I\t3500\t4000\t.\t-1.750\t.\t3500\t4000\t230,230,230\n",
      "I\t4000\t4500\t.\t-3.325\t.\t4000\t4500\t29,140,190\n",
      "I\t4500\t5000\t.\t-2.628\t.\t4500\t5000\t52,168,194\n",
      "(pycoMeth) chromosome\tstart\tend\tnum_motifs\tmedian_llr\tllr_list\n",
      "I\t500\t1000\t12\t-3.35\t[-1.14,-3.54,-7.24,-4.3,0.56,-0.65,-4.37,-3.78,-0.27,-1.32,-3.35]\n",
      "I\t1000\t1500\t22\t-3.65\t[-2.48,-5.035,-4.16,-3.315,-3.295,-1.69,-9.885,-7.95,-3.65,-3.0,-2.83,-8.36,-8.56]\n",
      "I\t1500\t2000\t19\t-3.4\t[-5.71,-6.05,-0.925,-7.165,-3.975,0.56,-1.78,-1.86,-3.67,-3.4,-5.53,-1.06,-1.79,-1.94,-6.22]\n",
      "I\t2000\t2500\t15\t-4.272\t[-5.24,-3.07,-4.33,-19.055,-7.55,-1.255,-2.565,-4.215,-5.515,-3.48]\n",
      "I\t2500\t3000\t19\t-2.5\t[-0.705,0.385,-6.685,-10.175,-4.27,-2.3,-2.5,-2.52,-1.425,-4.255,-2.73,0.35,-9.31,-1.05,-0.99]\n",
      "I\t3000\t3500\t9\t-1.4\t[-2.34,-2.19,-1.4,-1.0,0.29,-1.3,-6.9,-1.22,-1.92]\n",
      "I\t3500\t4000\t8\t-1.75\t[-1.7,-0.53,-4.46,-2.2,-1.75,-8.47,0.53]\n",
      "I\t4000\t4500\t10\t-3.325\t[-2.83,-3.96,-1.77,-13.895,-3.82,-1.73]\n",
      "I\t4500\t5000\t12\t-2.628\t[-2.71,-0.76,-1.55,-2.56,-0.985,-5.02,-3.46,-1.275,-2.695,-3.52]\n",
      "(pycoMeth) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoMeth Interval_Aggregate\\\n",
    "    -i ./data/CpG_Aggregate_sample_1.tsv \\\n",
    "    -f ./data/ref.fa \\\n",
    "    -b ./results/Interval_Aggregate_sample_1_CLI.bed \\\n",
    "    -t ./results/Interval_Aggregate_sample_1_CLI.tsv \\\n",
    "    --interval_size 500 \\\n",
    "    --min_cpg_per_interval 5 \\\n",
    "    -s sample_1\n",
    "\n",
    "head ./results/Interval_Aggregate_sample_1_CLI.bed\n",
    "head ./results/Interval_Aggregate_sample_1_CLI.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Usage with an annotation Bed file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-26T12:53:03.135141Z",
     "start_time": "2020-02-26T12:53:00.667186Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "## Checking options and input files ##\n",
      "## Parsing CpG_aggregate file ##\n",
      "\tResults summary\n",
      "\t\tLines parsed: 89,235\n",
      "\t\tTotal number of intervals: 2,041\n",
      "\tWritter summary\n",
      "\t\tEmpty intervals skipped: 1,323\n",
      "\t\tValid intervals written: 637\n",
      "\t\tLow CpG intervals skipped: 81\n",
      "(pycoMeth) (pycoMeth) track name=sample_1_Interval itemRgb=On\n",
      "I\t1804\t2170\t.\t-3.670\t.\t1804\t2170\t29,140,190\n",
      "I\t31835\t32949\t.\t-5.650\t.\t31835\t32949\t35,70,156\n",
      "I\t33497\t34371\t.\t-3.295\t.\t33497\t34371\t29,140,190\n",
      "I\t44730\t44988\t.\t-3.200\t.\t44730\t44988\t29,140,190\n",
      "I\t47889\t48187\t.\t-4.550\t.\t47889\t48187\t33,102,171\n",
      "I\t57175\t57391\t.\t-4.760\t.\t57175\t57391\t33,102,171\n",
      "I\t59052\t59257\t.\t-4.050\t.\t59052\t59257\t33,102,171\n",
      "I\t60422\t60656\t.\t-3.615\t.\t60422\t60656\t29,140,190\n",
      "I\t61246\t61903\t.\t-3.423\t.\t61246\t61903\t29,140,190\n",
      "(pycoMeth) chromosome\tstart\tend\tnum_motifs\tmedian_llr\tllr_list\n",
      "I\t1804\t2170\t14\t-3.67\t[-3.67,-3.4,-5.53,-1.06,-1.79,-1.94,-6.22,-5.24,-3.07,-4.33,-19.055]\n",
      "I\t31835\t32949\t10\t-5.65\t[-2.925,-6.055,-1.785,-5.65,-6.83,-1.695,-12.32]\n",
      "I\t33497\t34371\t19\t-3.295\t[-4.38,-3.32,-1.29,-3.27,-5.89,-8.96,-6.88,-2.22,-3.605,-1.07,-6.465,-1.735,0.15,-0.96]\n",
      "I\t44730\t44988\t9\t-3.2\t[-2.37,-4.9,-1.63,-1.69,-8.09,-4.03]\n",
      "I\t47889\t48187\t13\t-4.55\t[-4.55,-9.41,-3.37,-4.66,-3.24,-4.66,-4.535]\n",
      "I\t57175\t57391\t9\t-4.76\t[-7.96,-4.76,-0.33,-3.77,-1.03,-7.68,-6.66]\n",
      "I\t59052\t59257\t13\t-4.05\t[-8.53,-1.23,-11.15,-3.07,-3.88,-4.22,-1.59,-5.79]\n",
      "I\t60422\t60656\t18\t-3.615\t[-3.3,-3.14,-0.44,-6.65,-3.93,-2.27,-9.16,-1.73,-17.33,-4.25]\n",
      "I\t61246\t61903\t38\t-3.423\t[-3.68,-26.15,-6.45,-0.76,-3.86,-2.33,-4.945,-4.975,-2.615,-2.715,-2.905,-3.415,-9.04,-21.33,-2.735,-1.355,-3.43,-2.44]\n",
      "(pycoMeth) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoMeth Interval_Aggregate\\\n",
    "    -i ./data/CpG_Aggregate_sample_1.tsv \\\n",
    "    -f ./data/ref.fa \\\n",
    "    -a ./data/Yeast_CGI.bed \\\n",
    "    -b ./results/Interval_Aggregate_sample_1_interval_CLI.bed \\\n",
    "    -t ./results/Interval_Aggregate_sample_1_interval_CLI.tsv \\\n",
    "    -s sample_1\n",
    "\n",
    "head ./results/Interval_Aggregate_sample_1_interval_CLI.bed\n",
    "head ./results/Interval_Aggregate_sample_1_interval_CLI.tsv"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "hide_input": false,
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
