{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-08-10T10:03:00.595966Z",
     "start_time": "2018-08-10T10:03:00.587806Z"
    }
   },
   "source": [
    "# Meth_Comp API usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-07T14:11:46.223029Z",
     "start_time": "2020-04-07T14:11:46.216931Z"
    },
    "init_cell": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Import main module \n",
    "from pycoMeth.Meth_Comp import Meth_Comp\n",
    "\n",
    "# optionally inport jupyter helper functions\n",
    "from pycoMeth.common import head, jhelp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-07T14:11:47.817337Z",
     "start_time": "2020-04-07T14:11:47.769968Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**Meth_Comp** (aggregate_fn_list, ref_fasta_fn, output_tsv_fn, output_bed_fn, max_missing, min_diff_llr, sample_id_list, pvalue_adj_method, pvalue_threshold, verbose, quiet, progress, kwargs)\n",
       "\n",
       "Compare methylation values for each CpG positions or intervals between n samples and perform a statistical test to evaluate if the positions are significantly different. For 2 samples a Mann_Withney test is performed otherwise multiples samples are compared with a Kruskal Wallis test. pValues are adjusted for multiple tests using the Benjamini & Hochberg procedure for controlling the false discovery rate.\n",
       "\n",
       "---\n",
       "\n",
       "* **aggregate_fn_list** (required) [list(str)]\n",
       "\n",
       "A list of output tsv files corresponding to several samples to compare generated by either CpG_Aggregate or Interval_Aggregate. (can be gzipped)\n",
       "\n",
       "* **ref_fasta_fn** (required) [str]\n",
       "\n",
       "Reference file used for alignment in Fasta format (ideally already indexed with samtools faidx)\n",
       "\n",
       "* **output_tsv_fn** (default: None) [str]\n",
       "\n",
       "Path to write an more extensive result report in TSV format (At least 1 output file is required) (can be gzipped)\n",
       "\n",
       "* **output_bed_fn** (default: None) [str]\n",
       "\n",
       "Path to write a summary result file in BED format (At least 1 output file is required) (can be gzipped)\n",
       "\n",
       "* **max_missing** (default: 0) [int]\n",
       "\n",
       "Max number of missing samples to perform the test\n",
       "\n",
       "* **min_diff_llr** (default: 2) [float]\n",
       "\n",
       "Minimal llr boundary for negative and positive median llr. The test if only performed if at least one sample has a median llr above (methylated) and 1 sample has a median llr below (unmethylated)\n",
       "\n",
       "* **sample_id_list** (default: None) [list(str)]\n",
       "\n",
       "list of sample ids to annotate results in tsv file\n",
       "\n",
       "* **pvalue_adj_method** (default: fdr_bh) [str]\n",
       "\n",
       "Method to use for pValue multiple test adjustment\n",
       "\n",
       "* **pvalue_threshold** (default: 0.01) [float]\n",
       "\n",
       "Alpha parameter (family-wise error rate) for pValue adjustment\n",
       "\n",
       "* **verbose** (default: False) [bool]\n",
       "\n",
       "* **quiet** (default: False) [bool]\n",
       "\n",
       "* **progress** (default: False) [bool]\n",
       "\n",
       "* **kwargs**\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(Meth_Comp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Usage with CpG Aggregate output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-07T14:11:57.904171Z",
     "start_time": "2020-04-07T14:11:50.069516Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m## Checking options and input files ##\u001b[0m\n",
      "\u001b[01;34m## Parsing files ##\u001b[0m\n",
      "\u001b[32m\tReading input files header and checking consistancy between headers\u001b[0m\n",
      "\u001b[32m\tStarting asynchronous file parsing\u001b[0m\n",
      "\tProgress: 37.1M bytes [00:07, 4.80M bytes/s]                       \n",
      "\u001b[32m\tAdjust pvalues\u001b[0m\n",
      "\u001b[32m\tWriting output file\u001b[0m\n",
      "\tProgress: 100%|██████████| 344/344 [00:00<00:00, 20.0k sites/s]\n",
      "\u001b[32m\tResults summary\u001b[0m\n",
      "\u001b[32m\t\tSites with insufficient samples: 211,400\u001b[0m\n",
      "\u001b[32m\t\tSites with insufficient effect size: 33,071\u001b[0m\n",
      "\u001b[32m\t\tValid sites: 344\u001b[0m\n",
      "\u001b[32m\t\tSites with non-significant pvalue: 344\u001b[0m\n",
      "\u001b[32m\t\tSites with non-significant adjusted pvalue: 344\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chromosome start  end    n_samples pvalue              adj_pvalue          neg_med pos_med ambiguous_med labels                med_llr_list                raw_llr_list                                       \n",
      "I          118799 118800 4         0.29978058859571194 0.4279805197751711  3       1       0             [\"S1\",\"S2\",\"S3\",\"S4\"] [2.005,-2.645,-6.64,-5.51]  [[-1.82,5.83],[-0.09,-5.2],[-2.67,-10.61],[-6.7...\n",
      "I          141415 141416 4         0.11800901597381579 0.3742064977444858  1       1       2             [\"S1\",\"S2\",\"S3\",\"S4\"] [-4.175,0.135,3.02,-0.02]   [[-1.33,-4.52,-3.83,-4.67],[-1.42,1.69],[6.62,-...\n",
      "I          151819 151820 4         0.1315327225513005  0.3742064977444858  1       1       2             [\"S1\",\"S2\",\"S3\",\"S4\"] [-2.98,-1.95,2.93,-1.48]    [[-3.68,-4.4,0.54,-2.28],[-3.45,-1.95,-1.21],[4...\n",
      "I          167230 167237 4         0.10985439586067039 0.3742064977444858  3       1       0             [\"S1\",\"S2\",\"S3\",\"S4\"] [-6.305,-6.94,3.6,-4.17]    [[-5.06,-7.55],[-6.94,-7.26,-1.12],[3.6,5.68,0....\n",
      "II         64407  64408  4         0.16469672531155308 0.3742064977444858  1       1       2             [\"S1\",\"S2\",\"S3\",\"S4\"] [-1.06,-0.19,-3.68,2.93]    [[-1.06,-0.81,-2.26],[1.65,-2.03],[-3.85,-2.85,...\n",
      "II         102416 102417 4         0.06267205003507882 0.3742064977444858  2       1       1             [\"S1\",\"S2\",\"S3\",\"S4\"] [0.715,-3.05,2.17,-5.535]   [[-0.21,1.64],[-3.3,-2.8],[2.24,2.1],[-4.52,-2....\n",
      "II         108834 108835 4         0.1633613146761696  0.3742064977444858  3       1       0             [\"S1\",\"S2\",\"S3\",\"S4\"] [2.95,-5.16,-3.53,-4.53]    [[4.7,1.2],[-5.89,-4.43],[-6.55,-0.25,-3.53],[-...\n",
      "II         110457 110458 4         0.04436795794391088 0.3742064977444858  2       1       1             [\"S1\",\"S2\",\"S3\",\"S4\"] [-1.405,2.69,-3.285,-7.045] [[-0.97,-1.34,-1.47,-3.97],[-0.84,6.22],[-7.1,0...\n",
      "II         112989 112990 4         0.3618050274975317  0.46268003516412975 1       1       2             [\"S1\",\"S2\",\"S3\",\"S4\"] [-1.38,2.1,-1.94,-2.93]     [[-1.38,4.34,-5.48,-5.45,-0.93],[3.4,0.8],[-1.3...\n",
      "\n",
      "track name=meth_comp itemRgb=On\n",
      "I\t118799\t118800\t.\t0\t.\t118799\t118800\t230,230,230\n",
      "I\t141415\t141416\t.\t0\t.\t141415\t141416\t230,230,230\n",
      "I\t151819\t151820\t.\t0\t.\t151819\t151820\t230,230,230\n",
      "I\t167230\t167237\t.\t0\t.\t167230\t167237\t230,230,230\n",
      "II\t64407\t64408\t.\t0\t.\t64407\t64408\t230,230,230\n",
      "II\t102416\t102417\t.\t0\t.\t102416\t102417\t230,230,230\n",
      "II\t108834\t108835\t.\t0\t.\t108834\t108835\t230,230,230\n",
      "II\t110457\t110458\t.\t0\t.\t110457\t110458\t230,230,230\n",
      "II\t112989\t112990\t.\t0\t.\t112989\t112990\t230,230,230\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ff = Meth_Comp (\n",
    "    aggregate_fn_list=[\n",
    "        \"./data/Yeast_CpG_1.tsv.gz\", \n",
    "        \"./data/Yeast_CpG_2.tsv.gz\", \n",
    "        \"./data/Yeast_CpG_3.tsv.gz\", \n",
    "        \"./data/Yeast_CpG_4.tsv.gz\"],\n",
    "    ref_fasta_fn=\"./data/yeast.fa\",\n",
    "    output_bed_fn=\"./results/Yeast_CpG_meth_comp.bed\",\n",
    "    output_tsv_fn=\"./results/Yeast_CpG_meth_comp.tsv.gz\",\n",
    "    sample_id_list=[\"S1\",\"S2\",\"S3\",\"S4\"],\n",
    "    progress=True)\n",
    "\n",
    "head(\"./results/Yeast_CpG_meth_comp.tsv.gz\")\n",
    "head(\"./results/Yeast_CpG_meth_comp.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Usage with Interval Aggregate output with a single significant result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-07T14:12:04.422952Z",
     "start_time": "2020-04-07T14:12:04.270710Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m## Checking options and input files ##\u001b[0m\n",
      "\u001b[01;34m## Parsing files ##\u001b[0m\n",
      "\u001b[32m\tReading input files header and checking consistancy between headers\u001b[0m\n",
      "\u001b[32m\tStarting asynchronous file parsing\u001b[0m\n",
      "\tProgress: 776k bytes [00:00, 7.05M bytes/s]              \n",
      "\u001b[32m\tAdjust pvalues\u001b[0m\n",
      "\u001b[32m\tWriting output file\u001b[0m\n",
      "\tProgress: 100%|██████████| 4.00/4.00 [00:00<00:00, 9.99k sites/s]\n",
      "\u001b[32m\tResults summary\u001b[0m\n",
      "\u001b[32m\t\tSites with insufficient effect size: 938\u001b[0m\n",
      "\u001b[32m\t\tSites with insufficient samples: 921\u001b[0m\n",
      "\u001b[32m\t\tValid sites: 4\u001b[0m\n",
      "\u001b[32m\t\tSites with non-significant adjusted pvalue: 4\u001b[0m\n",
      "\u001b[32m\t\tSites with non-significant pvalue: 3\u001b[0m\n",
      "\u001b[32m\t\tSites with significant pvalue: 1\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chromosome start   end     n_samples pvalue               adj_pvalue           neg_med pos_med ambiguous_med unique_cpg_pos labels                med_llr_list               raw_llr_list                                       raw_pos_list                                       \n",
      "V          65380   65612   3         0.007390942610433672 0.029563770441734687 2       1       0             8              [\"S1\",\"S3\",\"S4\"]      [-2.28,0.542,-2.125]       [[-2.515,-0.06,-3.84,-0.475,-3.905,-11.445,-2.0...[[65382,65416,65434,65470,65482,65497,65527,655...\n",
      "X          745486  745750  3         0.1931198217867195   0.257493095715626    2       1       0             3              [\"S1\",\"S3\",\"S4\"]      [-3.42,-2.475,0.28]        [[-3.65,-3.42,-0.755],[-2.475,-9.35,-1.88],[0.2...[[745492,745563,745592],[745492,745563,745592],...\n",
      "XV         108248  108543  3         0.7757164275739235   0.7757164275739235   2       1       0             9              [\"S2\",\"S3\",\"S4\"]      [0.425,-2.445,-2.77]       [[0.425,-8.775,1.17],[-4.515,-4.21,-3.645,-2.44...[[108482,108497,108514],[108288,108308,108319,1...\n",
      "XV         1082847 1083061 4         0.058554972459289256 0.11710994491857851  3       1       0             3              [\"S1\",\"S2\",\"S3\",\"S4\"] [-3.47,-2.49,-6.875,0.395] [[-1.99,-4.68,-3.47],[-2.49,-0.83,-6.38],[-6.87...[[1082929,1082947,1082962],[1082929,1082947,108...\n",
      "\n",
      "track name=meth_comp itemRgb=On\n",
      "V\t65380\t65612\t.\t1\t.\t65380\t65612\t230,230,230\n",
      "X\t745486\t745750\t.\t0\t.\t745486\t745750\t230,230,230\n",
      "XV\t108248\t108543\t.\t0\t.\t108248\t108543\t230,230,230\n",
      "XV\t1082847\t1083061\t.\t0\t.\t1082847\t1083061\t230,230,230\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ff = Meth_Comp (\n",
    "    aggregate_fn_list=[\n",
    "        \"./data/Yeast_CGI_1.tsv.gz\", \n",
    "        \"./data/Yeast_CGI_2.tsv.gz\", \n",
    "        \"./data/Yeast_CGI_3.tsv.gz\", \n",
    "        \"./data/Yeast_CGI_4.tsv.gz\"],\n",
    "    ref_fasta_fn=\"./data/yeast.fa\",\n",
    "    output_bed_fn=\"./results/Yeast_CGI_meth_comp.bed\",\n",
    "    output_tsv_fn=\"./results/Yeast_CGI_meth_comp.tsv.gz\",\n",
    "    sample_id_list=[\"S1\",\"S2\",\"S3\",\"S4\"],\n",
    "    max_missing = 1,\n",
    "    min_diff_llr = 0,\n",
    "    progress=True)\n",
    "\n",
    "head(\"./results/Yeast_CGI_meth_comp.tsv.gz\")\n",
    "head(\"./results/Yeast_CGI_meth_comp.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Usage with Interval Aggregate output and larger dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-07T14:13:00.965770Z",
     "start_time": "2020-04-07T14:12:08.072684Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m## Checking options and input files ##\u001b[0m\n",
      "\u001b[01;34m## Parsing files ##\u001b[0m\n",
      "\u001b[32m\tReading input files header and checking consistancy between headers\u001b[0m\n",
      "\u001b[32m\tStarting asynchronous file parsing\u001b[0m\n",
      "\tProgress: 556M bytes [00:49, 11.2M bytes/s]                       \n",
      "\u001b[32m\tAdjust pvalues\u001b[0m\n",
      "\u001b[32m\tWriting output file\u001b[0m\n",
      "\tProgress: 100%|██████████| 10.8k/10.8k [00:02<00:00, 3.60k sites/s]\n",
      "\u001b[32m\tResults summary\u001b[0m\n",
      "\u001b[32m\t\tSites with insufficient effect size: 201,365\u001b[0m\n",
      "\u001b[32m\t\tSites with insufficient samples: 54,136\u001b[0m\n",
      "\u001b[32m\t\tValid sites: 10,784\u001b[0m\n",
      "\u001b[32m\t\tSites with non-significant adjusted pvalue: 7,252\u001b[0m\n",
      "\u001b[32m\t\tSites with non-significant pvalue: 6,506\u001b[0m\n",
      "\u001b[32m\t\tSites with significant pvalue: 4,278\u001b[0m\n",
      "\u001b[32m\t\tSites with significant adjusted pvalue: 3,532\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chromosome start  end    n_samples pvalue                adj_pvalue             neg_med pos_med ambiguous_med unique_cpg_pos labels                                   med_llr_list                             raw_llr_list                             raw_pos_list                             \n",
      "1          136203 136504 12        0.7664606178319608    0.8325424217694418     1       5       6             9              [\"11-1_A3\",\"117-2_C4\",\"131-1_F4\",\"134...[0.46,0.47,0.66,1.6,1.14,1.01,0.18,0....[[1.19,-0.05,4.49,-1.22,-3.13,0.46,-1...[[136270,136281,136306,136348,136368,...\n",
      "1          308880 309181 12        0.0006937034588303196 0.002644361293752621   8       1       3             6              [\"11-1_A3\",\"117-2_C4\",\"131-1_F4\",\"134...[0.655,-1.875,0.57,-1.152,2.145,-2.92...[[-0.1,1.41,5.22,-1.51,-0.23,1.59],[-...[[308888,308972,309024,309052,309091,...\n",
      "1          416519 416885 11        0.4452430035616395    0.5407095214424235     2       3       6             12             [\"11-1_A3\",\"117-2_C4\",\"131-1_F4\",\"134...[-1.92,1.538,0.365,0.18,1.852,0.64,0....[[0.15,5.61,2.36,-2.21,2.75,2.54,-1.6...[[416564,416587,416630,416648,416661,...\n",
      "1          436047 436994 11        0.4055421456046574    0.5019357854011965     10      1       0             24             [\"11-1_A3\",\"117-2_C4\",\"131-1_F4\",\"134...[2.16,-3.915,-3.758,-3.605,-3.035,-2....[[1.065,0.865,11.17,1.68,5.38,2.6,2.1...[[436726,436775,436794,436877,436922,...\n",
      "1          438148 438361 11        0.39304056783348      0.4895529548990816     1       2       8             6              [\"11-1_A3\",\"117-2_C4\",\"131-1_F4\",\"134...[1.38,-0.25,1.645,-0.245,0.005,-1.368...[[2.57,-0.345,-1.545,3.67,0.83,1.93],...[[438162,438206,438232,438261,438278,...\n",
      "1          486249 486512 12        0.2533257680909372    0.3432137034720149     1       5       6             5              [\"11-1_A3\",\"117-2_C4\",\"131-1_F4\",\"134...[1.59,0.42,1.04,0.2,0.05,-1.53,-0.685...[[0.035,1.92,0.61,4.035,1.59],[-1.25,...[[486252,486317,486396,486435,486490]...\n",
      "1          503417 503927 11        4.930402378514754e-05 0.0002634760121402532  2       9       0             14             [\"11-1_A3\",\"117-2_C4\",\"131-1_F4\",\"134...[2.598,1.795,1.522,-1.142,-5.395,1.73...[[-0.86,-0.87,-2.935,1.895,1.215,5.05...[[503443,503497,503540,503564,503593,...\n",
      "1          514965 516003 12        6.942129597393029e-26 9.242459947936596e-24  4       7       1             32             [\"11-1_A3\",\"117-2_C4\",\"131-1_F4\",\"134...[-2.23,-3.225,1.2,1.265,3.05,0.505,3....[[2.48,1.3,4.55,2.79,2.04,2.98,0.84,-...[[514980,514998,515042,515096,515128,...\n",
      "1          570414 576486 11        7.321113498353957e-06 4.8645032634780694e-05 1       7       3             184            [\"117-2_C4\",\"131-1_F4\",\"134-1_H4\",\"13...[0.66,1.628,1.518,1.638,1.078,1.652,-...[[1.89,1.6,7.95,1.235,1.935,1.515,2.2...[[570501,570666,570699,570732,570765,...\n",
      "\n",
      "track name=meth_comp itemRgb=On\n",
      "1\t136203\t136504\t.\t0\t.\t136203\t136504\t230,230,230\n",
      "1\t308880\t309181\t.\t2\t.\t308880\t309181\t209,70,67\n",
      "1\t416519\t416885\t.\t0\t.\t416519\t416885\t230,230,230\n",
      "1\t436047\t436994\t.\t0\t.\t436047\t436994\t230,230,230\n",
      "1\t438148\t438361\t.\t0\t.\t438148\t438361\t230,230,230\n",
      "1\t486249\t486512\t.\t0\t.\t486249\t486512\t230,230,230\n",
      "1\t503417\t503927\t.\t3\t.\t503417\t503927\t187,55,84\n",
      "1\t514965\t516003\t.\t23\t.\t514965\t516003\t10,7,35\n",
      "1\t570414\t576486\t.\t4\t.\t570414\t576486\t163,43,97\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Generate list of file paths and sample ids from source directory\n",
    "from glob import glob\n",
    "import os\n",
    "fn_list = sorted(glob(\"./data/medaka_CGI_*\"))\n",
    "id_list = [os.path.split(fn)[-1].strip(\"medaka_CGI_\").strip(\".tsv.gz\") for fn in fn_list]\n",
    "\n",
    "Meth_Comp (\n",
    "    aggregate_fn_list=fn_list,\n",
    "    ref_fasta_fn=\"./data/medaka.fa\",\n",
    "    output_bed_fn=\"./results/Medaka_CGI_meth_comp.bed\",\n",
    "    output_tsv_fn=\"./results/Medaka_CGI_meth_comp.tsv.gz\",\n",
    "    sample_id_list = id_list,\n",
    "    max_missing = 1,\n",
    "    min_diff_llr = 1,\n",
    "    progress=True)\n",
    "\n",
    "head(\"./results/Medaka_CGI_meth_comp.tsv.gz\", max_char_col=40)\n",
    "head(\"./results/Medaka_CGI_meth_comp.bed\", max_char_col=40)"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "hide_input": false,
  "kernelspec": {
   "display_name": "pycoMeth",
   "language": "python",
   "name": "pycometh"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
