{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-08-10T10:03:00.595966Z",
     "start_time": "2018-08-10T10:03:00.587806Z"
    }
   },
   "source": [
    "# Meth_Comp CLI usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Activate virtual environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-14T23:13:49.653438Z",
     "start_time": "2020-01-14T23:13:48.621951Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(pycoMeth) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Using virtualenvwrapper here but can also be done with Conda \n",
    "workon pycoMeth"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-13T12:30:11.252204Z",
     "start_time": "2020-01-13T12:30:10.066234Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: pycoMeth Meth_Comp [-h] -i [AGGREGATE_FN_LIST [AGGREGATE_FN_LIST ...]]\n",
      "                          -f REF_FASTA_FN [-b OUTPUT_TSV_FN]\n",
      "                          [-t OUTPUT_BED_FN] [-m MAX_MISSING]\n",
      "                          [-l MIN_DIFF_LLR]\n",
      "                          [-s [SAMPLE_ID_LIST [SAMPLE_ID_LIST ...]]]\n",
      "                          [--pvalue_adj_method PVALUE_ADJ_METHOD]\n",
      "                          [--pvalue_adj_alpha PVALUE_ADJ_ALPHA] [-v] [-q] [-p]\n",
      "\n",
      "Compare methylation values for each CpG positions or intervals between n\n",
      "samples and perform a statistical test to evaluate if the positions are\n",
      "significantly different. For 2 samples a Mann_Withney test is performed\n",
      "otherwise multiples samples are compared with a Kruskal Wallis test. pValues\n",
      "are adjusted for multiple tests using the Benjamini & Hochberg procedure for\n",
      "controlling the false discovery rate.\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "\n",
      "Input/Output options:\n",
      "  -i [AGGREGATE_FN_LIST [AGGREGATE_FN_LIST ...]], --aggregate_fn_list [AGGREGATE_FN_LIST [AGGREGATE_FN_LIST ...]]\n",
      "                        A list of output tsv files corresponding to several\n",
      "                        samples to compare generated by either CpG_Aggregate\n",
      "                        or Interval_Aggregate. (can be gzipped) (required)\n",
      "                        [str]\n",
      "  -f REF_FASTA_FN, --ref_fasta_fn REF_FASTA_FN\n",
      "                        Reference file used for alignment in Fasta format\n",
      "                        (ideally already indexed with samtools faidx)\n",
      "                        (required) [str]\n",
      "  -b OUTPUT_TSV_FN, --output_tsv_fn OUTPUT_TSV_FN\n",
      "                        Path to write an more extensive result report in TSV\n",
      "                        format (At least 1 output file is required) (can be\n",
      "                        gzipped) (default: None) [str]\n",
      "  -t OUTPUT_BED_FN, --output_bed_fn OUTPUT_BED_FN\n",
      "                        Path to write a summary result file in BED format (At\n",
      "                        least 1 output file is required) (can be gzipped)\n",
      "                        (default: None) [str]\n",
      "\n",
      "Misc options:\n",
      "  -m MAX_MISSING, --max_missing MAX_MISSING\n",
      "                        Max number of missing samples to perform the test\n",
      "                        (default: 0) [int]\n",
      "  -l MIN_DIFF_LLR, --min_diff_llr MIN_DIFF_LLR\n",
      "                        Minimal llr boundary for negative and positive median\n",
      "                        llr. The test if only performed if at least one sample\n",
      "                        has a median llr above (methylated) and 1 sample has a\n",
      "                        median llr below (unmethylated) (default: 2) [float]\n",
      "  -s [SAMPLE_ID_LIST [SAMPLE_ID_LIST ...]], --sample_id_list [SAMPLE_ID_LIST [SAMPLE_ID_LIST ...]]\n",
      "                        list of sample ids to annotate results in tsv file\n",
      "                        (default: None) [str]\n",
      "  --pvalue_adj_method PVALUE_ADJ_METHOD\n",
      "                        Method to use for pValue multiple test adjustment\n",
      "                        (default: fdr_bh) [str]\n",
      "  --pvalue_adj_alpha PVALUE_ADJ_ALPHA\n",
      "                        Alpha parameter (family-wise error rate) for pValue\n",
      "                        adjustment (default: 0.01) [float]\n",
      "\n",
      "Verbosity options:\n",
      "  -v, --verbose         Increase verbosity\n",
      "  -q, --quiet           Reduce verbosity\n",
      "  -p, --progress        Display a progress bar\n",
      "(pycoMeth) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoMeth Meth_Comp --help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Usage with CpG Aggregate output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-13T12:32:43.371129Z",
     "start_time": "2020-01-13T12:32:33.067592Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "## Checking options and input files ##\n",
      "## Parsing files ##\n",
      "\tReading input files header and checking consistancy between headers\n",
      "\tStarting asynchronous file parsing\n",
      "37.1M bytes [00:08, 4.35M bytes/s]                                              \n",
      "\tAdjust pvalues\n",
      "\tWriting output file\n",
      "100%|█████████████████████████████████| 3.28k/3.28k [00:00<00:00, 26.7k sites/s]\n",
      "\tResults summary\n",
      "\t\tSites with insufficient samples: 211,400\n",
      "\t\tSites with insufficient effect size: 30,132\n",
      "\t\tValid sites: 3,283\n",
      "\t\tSites with significant pvalue (< 0.01): 124\n",
      "(pycoMeth) (pycoMeth) chromosome\tstart\tend\tn_samples\tpvalue\tstatistic\tadj_pvalue\tneg_med\tpos_med\tambiguous_med\tlabels\tmed_llr_list\traw_llr_list\n",
      "I\t109610\t109611\t4\t0.5355384957105982\t2.18181818181818\t0.637278314240832\t3\t1\t0\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[1.06,-3.0,-3.285,-3.575]\t[[2.83,-0.71],[-4.85,1.83,-3.0],[0.87,-7.44],[0.13,-4.22,-2.93,-6.8]]\n",
      "I\t109667\t109668\t4\t0.6210918363528521\t1.7717948717948744\t0.6997407339555297\t3\t1\t0\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[0.555,-1.45,-4.96,-2.57]\t[[5.93,-4.82],[-1.45,-1.28,-5.4],[-5.79,-4.13],[-5.81,-1.87,-0.82,-2.57,-5.08]]\n",
      "I\t110046\t110047\t4\t0.4846302240896865\t2.448717948717942\t0.6031859226921511\t2\t1\t1\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[0.3,-1.33,-3.355,1.32]\t[[-2.0,2.6],[-1.94,-1.33,-1.2],[-1.73,-4.98],[3.33,1.8,-1.03,-5.67,1.32]]\n",
      "I\t110157\t110158\t4\t0.10079374072306878\t6.233333333333334\t0.4770356012343023\t3\t1\t0\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[-0.535,1.14,-3.28,-2.96]\t[[1.03,-2.1],[1.14,3.45,-0.49],[-2.79,-3.77],[-3.61,-0.77,-2.96,-0.32,-7.67]]\n",
      "I\t110226\t110227\t4\t0.2562512982158426\t4.04871794871795\t0.48938810639848035\t3\t1\t0\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[1.025,-1.21,-0.875,-2.85]\t[[1.13,0.92],[0.46,-1.45,-1.21],[-1.62,-0.13],[-6.69,2.96,-4.64,-1.5,-2.85]]\n",
      "I\t110246\t110247\t4\t0.16789646023633203\t5.053846153846159\t0.4770356012343023\t2\t1\t1\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[-2.63,0.57,-2.05,0.2]\t[[-3.84,-1.42],[-1.04,0.91,0.57],[-3.46,-0.64],[-3.17,0.17,0.94,0.2,1.83]]\n",
      "I\t110689\t110690\t4\t0.15407920997458144\t5.2545454545454575\t0.4770356012343023\t3\t1\t0\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[-1.885,0.5,-6.53,-2.37]\t[[-1.89,-1.88],[1.32,-0.32],[-2.53,-10.53],[-1.76,0.49,-5.91,-2.37,-4.05]]\n",
      "I\t110925\t110926\t4\t0.45883970397306584\t2.592307692307699\t0.5920104435099945\t3\t1\t0\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[-3.06,-5.365,-1.785,0.61]\t[[-3.06,-4.18,2.33],[-10.58,-0.15],[-5.51,1.94],[1.84,-1.45,0.61,2.65,0.35]]\n",
      "I\t111021\t111022\t4\t0.10926389869102016\t6.04871794871795\t0.4770356012343023\t3\t1\t0\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[-7.8,-4.85,-4.94,0.51]\t[[-7.8,-3.83,-8.36],[-3.64,-6.06],[-4.19,-5.69],[2.71,1.03,0.32,-5.6,0.51]]\n",
      "(pycoMeth) track name=meth_comp itemRgb=On\n",
      "I\t109610\t109611\t.\t0.196\t.\t109610\t109611\t'230,230,230'\n",
      "I\t109667\t109668\t.\t0.155\t.\t109667\t109668\t'230,230,230'\n",
      "I\t110046\t110047\t.\t0.22\t.\t110046\t110047\t'230,230,230'\n",
      "I\t110157\t110158\t.\t0.321\t.\t110157\t110158\t'230,230,230'\n",
      "I\t110226\t110227\t.\t0.31\t.\t110226\t110227\t'230,230,230'\n",
      "I\t110246\t110247\t.\t0.321\t.\t110246\t110247\t'230,230,230'\n",
      "I\t110689\t110690\t.\t0.321\t.\t110689\t110690\t'230,230,230'\n",
      "I\t110925\t110926\t.\t0.228\t.\t110925\t110926\t'230,230,230'\n",
      "I\t111021\t111022\t.\t0.321\t.\t111021\t111022\t'230,230,230'\n",
      "(pycoMeth) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoMeth Meth_Comp \\\n",
    "    -i \"./data/Yeast_sample_1_CpG.tsv.gz\" \"./data/Yeast_sample_2_CpG.tsv.gz\" \"./data/Yeast_sample_3_CpG.tsv.gz\" \"./data/Yeast_sample_4_CpG.tsv.gz\" \\\n",
    "    -f \"./data/yeast.fa\" \\\n",
    "    -b \"./results/CLI_CpG_Yeast.bed\" \\\n",
    "    -t \"./results/CLI_CpG_Yeast.tsv\" \\\n",
    "    -s S1 S2 S3 S4 \\\n",
    "    -l 0.5 \\\n",
    "    --progress\n",
    "\n",
    "head ./results/CLI_CpG_Yeast.bed\n",
    "head ./results/CLI_CpG_Yeast.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Usage with Interval Aggregate output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-13T12:34:10.555147Z",
     "start_time": "2020-01-13T12:33:58.465907Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "## Checking options and input files ##\n",
      "## Parsing files ##\n",
      "\tReading input files header and checking consistancy between headers\n",
      "\tStarting asynchronous file parsing\n",
      "91.7M bytes [00:10, 8.78M bytes/s]                                              \n",
      "\tAdjust pvalues\n",
      "\tWriting output file\n",
      "100%|█████████████████████████████████████| 636/636 [00:00<00:00, 18.7k sites/s]\n",
      "\tResults summary\n",
      "\t\tSites with insufficient samples: 35,530\n",
      "\t\tSites with insufficient effect size: 224,256\n",
      "\t\tValid sites: 636\n",
      "\t\tSites with significant pvalue (< 0.01): 592\n",
      "\t\tSites with significant FDR adj pvalue (< 0.01): 588\n",
      "(pycoMeth) (pycoMeth) chromosome\tstart\tend\tn_samples\tpvalue\tstatistic\tadj_pvalue\tneg_med\tpos_med\tambiguous_med\tlabels\tmed_llr_list\traw_llr_list\n",
      "1\t514965\t516003\t4\t4.843871813601806e-08\t36.894573449995676\t2.2652224069490797e-07\t1\t2\t1\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[0.505,2.88,-3.49,3.02]\t[[4.05,2.09,3.61,3.67,2.89,2.82,1.41,3.87,2.18,-2.58,-0.52,0.41,-0.25,0.42,-0.93,-3.32,1.73,0.59,-1.55,-0.23,-3.54,3.42,1.86,1.04,-0.88,-0.06,3.87,2.1,0.34,-0.04,-3.78,0.255],[2.6,1.6,1.93,4.24,1.86,3.35,2.68,7.51,3.81,-0.24,7.42,1.17,1.14,8.74,4.45,12.355,2.44,17.08,13.59,0.26,0.11,2.88,3.14,13.11,3.53,1.58,12.28,9.425,-2.51,-0.59,-1.16],[5.24,3.15,3.54,4.295,1.06,4.17,1.02,-4.85,-0.64,-9.38,1.96,-0.95,-1.17,-0.47,-4.69,1.455,-3.49,-11.4,-21.23,-4.99,-7.855,-1.395,-7.755,-11.41,-3.78,-3.595,-19.425,-18.98,-2.41,-4.455,-4.05],[2.79,0.85,2.65,3.41,1.79,3.14,2.97,8.13,3.93,-1.37,3.475,0.36,1.29,6.88,4.44,16.22,3.4,17.19,11.78,0.21,0.77,3.53,2.47,13.12,3.07,0.85,11.525,9.33,1.985,0.45,-2.87,0.69]]\n",
      "1\t727090\t729031\t4\t1.244793813160041e-10\t49.09550636661502\t1.3593811986963564e-09\t1\t3\t0\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[2.37,-3.302,2.6,2.082]\t[[1.01,1.01,1.375,2.33,3.01,1.3,2.8,1.95,2.29,6.86,3.115,4.695,1.52,4.88,3.33,3.18,2.895,1.535,3.69,0.845,0.0,7.76,0.95,4.46,8.04,4.62,4.58,0.06,1.84,2.06,1.5,1.51,1.57,2.71,2.31,2.27,6.42,2.1,9.34,4.84,11.72,5.72,3.01,2.68,1.875,1.36,0.41,5.0,3.2,0.715,1.44,6.43,0.595,5.85,9.46,3.58,-0.21,2.46,-0.38,2.41,0.82,1.09],[-0.7,-3.265,-3.52,-11.18,-4.26,-3.34,-1.32,-13.11,-2.81,-12.23,-7.85,-3.93,-0.48,-1.88,-2.83,-0.79,-0.67,-4.49],[1.15,0.33,0.215,0.865,2.895,1.6,3.435,2.895,1.035,8.39,1.89,4.23,2.84,7.83,2.07,2.55,2.65,3.98,1.51,-0.25,-0.24,6.98,-0.53,1.72,9.975,3.45,4.765,5.25,-0.34,3.3,0.86,1.23,0.71,2.34,1.77,3.17,8.85,0.01,11.7,2.68,11.34,7.88,3.42,3.08,2.31,0.55,0.53,4.37,1.59,0.87,3.795,7.365,2.405,10.825,11.04,2.65,0.76,3.11,0.41,3.73,0.72,0.705],[1.755,1.515,2.105,1.56,5.105,1.345,3.965,0.88,1.34,5.06,2.39,5.14,1.25,4.91,3.77,2.42,4.06,1.68,5.89,0.235,0.16,8.8,0.51,5.66,6.075,5.74,0.6,-0.21,1.27,2.975,0.42,0.82,1.28,3.435,2.105,1.2,6.095,1.655,7.63,4.44,12.065,6.835,2.965,2.215,1.39,1.645,-0.25,2.55,2.96,0.125,1.58,3.925,0.195,5.43,8.34,1.87,-0.94,2.33,-0.31,2.06,0.79,1.97]]\n",
      "1\t729239\t730182\t4\t2.3772399552637456e-09\t43.07083111453157\t1.8438105018874905e-08\t1\t1\t2\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[-1.055,-3.225,2.105,-0.225]\t[[-0.85,-5.86,-1.06,-1.59,0.13,-0.805,-3.9,-2.52,-1.895,-2.205,-9.285,-3.42,-0.815,-3.455,-0.635,-12.405,-2.38,-2.1,-4.285,-3.18,0.095,-2.65,-0.33,3.08,-0.22,0.605,1.43,-0.47,-1.05,-2.295,2.04,1.855,0.45,0.24],[-0.835,-9.625,-5.12,-1.39,-0.665,-2.425,-4.72,-3.11,-5.42,-9.96,-13.15,-6.47,-1.81,-4.24,-1.99,-14.01,-1.97,-3.2,-7.845,-5.73,-1.745,-5.45,-1.44,-8.46,-2.45,-3.25,0.255,-3.975,-1.87,-4.17,-0.165,-4.05,-0.31,-0.355],[2.015,9.87,3.645,0.315,2.195,1.035,3.385,1.935],[-0.24,0.71,1.64,-0.28,0.45,-0.41,-0.47,-0.21,1.77,1.005,-0.265,-3.01,-0.13,-0.04,-0.05,-5.04,-1.76,-0.83,0.31,-1.885,1.045,-2.82,-0.06,1.69,0.38,-0.57,-0.67,-2.95,-2.86,-1.89,0.07,-1.27,0.14,0.14]]\n",
      "1\t2610894\t2611179\t4\t0.00021663864601670918\t19.488488488488485\t0.00036450311869478053\t1\t2\t1\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[1.155,2.605,-4.47,3.93]\t[[1.155,0.3,5.785,-0.825,1.18,2.67,-3.26,1.58,-5.15],[2.605,1.845,5.25,-0.5,-6.91,4.15,1.25,3.145,15.025],[-2.93,-3.45,-14.45,-5.55,-8.67,-2.22,-4.47,-1.92,-18.14],[3.99,2.3,11.45,0.98,10.92,3.365,1.56,3.93,18.325]]\n",
      "1\t2611213\t2611524\t4\t4.1412221602355744e-06\t27.728495027209604\t1.1501385562925002e-05\t1\t2\t1\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[0.5,3.53,-3.985,3.35]\t[[2.07,-0.43,1.17,0.33,1.78,-0.46,0.31,0.67,2.24,-1.32],[7.08,3.2,1.5,1.98,4.93,0.71,5.1,1.27,3.86,4.65],[-6.195,-8.87,-2.185,-0.655,-6.16,-2.37,-13.27,-3.61,-1.75,-4.36],[5.5,4.38,4.61,2.17,4.83,1.17,2.49,2.17,4.21,-0.58]]\n",
      "1\t3739186\t3739861\t4\t1.6382877864048344e-05\t24.87693421183988\t3.844837757023892e-05\t2\t1\t1\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[-3.932,-1.53,-5.322,2.145]\t[[-0.87,-3.85,0.13,-5.55,-3.345,-4.89,-11.685,-4.015,-7.055,-0.925,-17.495,-1.96],[-0.085,-2.73,1.15,-2.92,-0.675,-0.21,-11.655,-3.465,-1.375,-1.075,-17.115,-7.03,0.145,-5.555,-3.125,0.85,-3.81,-1.71,-1.53,2.235,-0.96],[-3.045,-5.35,-1.38,-13.68,-4.25,-5.295,-18.08,-6.655,-12.46,-3.1,-23.93,-1.735],[3.77,4.16,2.09,3.45,2.2,0.75,1.65,0.58]]\n",
      "1\t3945291\t3945604\t4\t0.020451556682244607\t9.78857142857143\t0.02298090114824659\t1\t3\t0\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[5.87,4.95,4.415,-3.275]\t[[18.94,3.42,6.55,5.87,3.26],[17.52,1.405,5.98,4.95,0.835],[17.945,2.895,5.615,4.415,2.06],[-5.89,-2.855,-3.495,2.13,-3.275]]\n",
      "1\t4095356\t4095859\t4\t6.210027210724236e-06\t26.889455782312922\t1.6538998364044193e-05\t1\t1\t2\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[1.115,-2.35,2.062,1.88]\t[[0.745,0.2,0.635,1.21,2.92,1.02,-0.32,2.59,4.84,8.965,0.79,1.31],[-0.22,-0.835,-0.84,-1.07,-8.89,-9.93,-3.945,-1.48,-15.235,-16.12,-3.105,-1.595],[2.22,2.32,0.47,2.11,2.555,4.005,0.135,2.015,1.11,9.155,0.565,1.93],[1.59,0.98,2.39,1.18,4.115,3.99,-0.145,1.795,4.135,5.695,1.82,1.94]]\n",
      "1\t4117891\t4118603\t4\t2.9727245218414944e-08\t37.89650919648791\t1.5125222367129523e-07\t1\t3\t0\t[\"S1\",\"S2\",\"S3\",\"S4\"]\t[2.475,2.255,-3.77,2.38]\t[[3.825,8.31,1.675,2.475,1.065,-2.4,5.16,0.42,2.985,4.165,3.935,2.04,2.005,6.025,17.84,3.83,2.3,2.3,1.2,0.315,4.475],[6.75,7.595,1.21,1.095,0.785,-0.335,2.255,1.12,1.315,5.49,6.535,1.655,4.785,6.57,21.03,3.64,4.0,2.49,0.385,0.99,1.82],[2.185,2.06,0.285,-3.77,-1.935,-3.415,-10.08,-2.655,-4.405,-10.785,-5.165,-3.005,-2.335,-10.03,-31.89,-1.25,-3.84,-4.61,-3.81,-2.28,-4.98],[3.28,6.305,3.525,2.365,1.395,-3.34,-0.515,2.38,2.07,4.11,2.79,3.0,2.38,1.9,15.435,3.075,1.665,1.82,0.6,1.28,3.9]]\n",
      "(pycoMeth) track name=meth_comp itemRgb=On\n",
      "1\t514965\t516003\t.\t6.64\t.\t514965\t516003\t'112,25,110'\n",
      "1\t727090\t729031\t.\t8.87\t.\t727090\t729031\t'60,9,101'\n",
      "1\t729239\t730182\t.\t7.73\t.\t729239\t730182\t'87,15,109'\n",
      "1\t2610894\t2611179\t.\t3.44\t.\t2610894\t2611179\t'187,55,84'\n",
      "1\t2611213\t2611524\t.\t4.94\t.\t2611213\t2611524\t'163,43,97'\n",
      "1\t3739186\t3739861\t.\t4.42\t.\t3739186\t3739861\t'163,43,97'\n",
      "1\t3945291\t3945604\t.\t1.64\t.\t3945291\t3945604\t'228,90,49'\n",
      "1\t4095356\t4095859\t.\t4.78\t.\t4095356\t4095859\t'163,43,97'\n",
      "1\t4117891\t4118603\t.\t6.82\t.\t4117891\t4118603\t'112,25,110'\n",
      "(pycoMeth) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoMeth Meth_Comp \\\n",
    "    -i \"./data/Medaka_sample_1_CGI.tsv.gz\" \"./data/Medaka_sample_2_CGI.tsv.gz\" \"./data/Medaka_sample_3_CGI.tsv.gz\" \"./data/Medaka_sample_4_CGI.tsv.gz\" \\\n",
    "    -f \"./data/medaka.fa\" \\\n",
    "    -b \"./results/CLI_Interval_Medaka.bed\" \\\n",
    "    -t \"./results/CLI_Interval_Medaka.tsv\" \\\n",
    "    -s S1 S2 S3 S4 \\\n",
    "    --progress\n",
    "\n",
    "head ./results/CLI_Interval_Medaka.bed\n",
    "head ./results/CLI_Interval_Medaka.tsv"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "hide_input": false,
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
