{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-08-10T10:03:00.595966Z",
     "start_time": "2018-08-10T10:03:00.587806Z"
    }
   },
   "source": [
    "# Meth_Comp CLI usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Activate virtual environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-08T14:42:59.417804Z",
     "start_time": "2020-01-08T14:42:58.592544Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(pycoMeth) (pycoMeth) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Using virtualenvwrapper here but can also be done with Conda \n",
    "workon pycoMeth"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-08T14:43:01.106391Z",
     "start_time": "2020-01-08T14:42:59.862282Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: pycoMeth Meth_Comp [-h] -i [AGGREGATE_FN_LIST [AGGREGATE_FN_LIST ...]]\n",
      "                          -f REF_FASTA_FN [-b OUTPUT_TSV_FN]\n",
      "                          [-t OUTPUT_BED_FN] [-m MAX_MISSING]\n",
      "                          [-l MIN_DIFF_LLR]\n",
      "                          [-s [SAMPLE_ID_LIST [SAMPLE_ID_LIST ...]]]\n",
      "                          [--pvalue_adj_method PVALUE_ADJ_METHOD]\n",
      "                          [--pvalue_adj_alpha PVALUE_ADJ_ALPHA] [-v] [-q] [-p]\n",
      "\n",
      "Compare methylation values for each CpG positions or intervals between n\n",
      "samples and perform a statistical test to evaluate if the positions are\n",
      "significantly different. For 2 samples a Mann_Withney test is performed\n",
      "otherwise multiples samples are compared with a Kruskal Wallis test. pValues\n",
      "are adjusted for multiple tests using the Benjamini & Hochberg procedure for\n",
      "controlling the false discovery rate.\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "\n",
      "Input/Output options:\n",
      "  -i [AGGREGATE_FN_LIST [AGGREGATE_FN_LIST ...]], --aggregate_fn_list [AGGREGATE_FN_LIST [AGGREGATE_FN_LIST ...]]\n",
      "                        A list of output tsv files corresponding to several\n",
      "                        samples to compare generated by either CpG_Aggregate\n",
      "                        or Interval_Aggregate. (required) [str]\n",
      "  -f REF_FASTA_FN, --ref_fasta_fn REF_FASTA_FN\n",
      "                        Reference file used for alignment in Fasta format\n",
      "                        (ideally already indexed with samtools faidx)\n",
      "                        (required) [str]\n",
      "  -b OUTPUT_TSV_FN, --output_tsv_fn OUTPUT_TSV_FN\n",
      "                        Path to write an more extensive result report in TSV\n",
      "                        format (At least 1 output file is required) (default:\n",
      "                        None) [str]\n",
      "  -t OUTPUT_BED_FN, --output_bed_fn OUTPUT_BED_FN\n",
      "                        Path to write a summary result file in BED format (At\n",
      "                        least 1 output file is required) (default: None) [str]\n",
      "\n",
      "Misc options:\n",
      "  -m MAX_MISSING, --max_missing MAX_MISSING\n",
      "                        Max number of missing samples to perform the test\n",
      "                        (default: 0) [int]\n",
      "  -l MIN_DIFF_LLR, --min_diff_llr MIN_DIFF_LLR\n",
      "                        Minimal llr boundary for negative and positive median\n",
      "                        llr. The test if only performed if at least one sample\n",
      "                        has a median llr above (methylated) and 1 sample has a\n",
      "                        median llr below (unmethylated) (default: 2) [float]\n",
      "  -s [SAMPLE_ID_LIST [SAMPLE_ID_LIST ...]], --sample_id_list [SAMPLE_ID_LIST [SAMPLE_ID_LIST ...]]\n",
      "                        list of sample ids to annotate results in tsv file\n",
      "                        (default: None) [str]\n",
      "  --pvalue_adj_method PVALUE_ADJ_METHOD\n",
      "                        Method to use for pValue multiple test adjustment\n",
      "                        (default: fdr_bh) [str]\n",
      "  --pvalue_adj_alpha PVALUE_ADJ_ALPHA\n",
      "                        Alpha parameter (family-wise error rate) for pValue\n",
      "                        adjustment (default: 0.01) [float]\n",
      "\n",
      "Verbosity options:\n",
      "  -v, --verbose         Increase verbosity\n",
      "  -q, --quiet           Reduce verbosity\n",
      "  -p, --progress        Display a progress bar\n",
      "(pycoMeth) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoMeth Meth_Comp --help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Usage with CpG Aggregate output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-07T10:42:01.533849Z",
     "start_time": "2020-01-07T10:41:57.012997Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "## Checking options and input files ##\n",
      "## Parsing files ##\n",
      "\tReading input files header and checking consistancy between headers\n",
      "\tStarting asynchronous file parsing\n",
      "\t: 100%|██████████████████████████████| 16.8M/16.8M [00:02<00:00, 6.22M bytes/s]\n",
      "\tAdjust pvalues\n",
      "\tWriting output file\n",
      "\t: 100%|████████████████████████████████| 3.00/3.00 [00:00<00:00, 13.5k sites/s]\n",
      "\tResults summary\n",
      "\t\tSites with insufficient samples: 182,480\n",
      "\t\tSites with insufficient effect size: 220\n",
      "\t\tValid sites: 3\n",
      "(pycoMeth) (pycoMeth) chromosome\tstart\tend\tn_samples\tpvalue\tstatistic\tadj_pvalue\tllr_list\n",
      "VIII\t212581\t212582\t4\t0.09220210690491402\t6.4364654952890135\t0.2766063207147421\t0.075;0.87;-2.3;-3.8\n",
      "VIII\t212710\t212711\t4\t0.42395933696735744\t2.7972104231918884\t0.42395933696735744\t-3.38;1.87;-4.265;-3.79\n",
      "VIII\t215884\t215885\t4\t0.23340395902747246\t4.27351472057353\t0.3501059385412087\t-4.61;0.73;-1.57;-3.525\n",
      "(pycoMeth) track name=CpG_Yeast_meth_comp itemRgb=On\n",
      "VIII\t212581\t212582\t.\t0.558\t.\t212581\t212582\t'100,100,100'\n",
      "VIII\t212710\t212711\t.\t0.373\t.\t212710\t212711\t'100,100,100'\n",
      "VIII\t215884\t215885\t.\t0.456\t.\t215884\t215885\t'100,100,100'\n",
      "(pycoMeth) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoMeth Meth_Comp \\\n",
    "    -i ./data/CpG_Aggregate_sample_*.tsv \\\n",
    "    -f ./data/yeast.fa \\\n",
    "    -b ./results/CLI_CpG_Yeast.bed \\\n",
    "    -t ./results/CLI_CpG_Yeast.tsv \\\n",
    "    -s CpG_Yeast \\\n",
    "    -l 0.5 \\\n",
    "    --progress\n",
    "\n",
    "head ./results/CLI_CpG_Yeast.bed\n",
    "head ./results/CLI_CpG_Yeast.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Usage with Interval Aggregate output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-07T10:43:44.478057Z",
     "start_time": "2020-01-07T10:43:19.541657Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "## Checking options and input files ##\n",
      "## Parsing files ##\n",
      "\tReading input files header and checking consistancy between headers\n",
      "\tStarting asynchronous file parsing\n",
      "\t: 100%|████████████████████████████████| 322M/322M [00:23<00:00, 13.8M bytes/s]\n",
      "\tAdjust pvalues\n",
      "\tWriting output file\n",
      "\t: 100%|██████████████████████████████████| 580/580 [00:00<00:00, 43.8k sites/s]\n",
      "\tResults summary\n",
      "\t\tSites with insufficient samples: 167,919\n",
      "\t\tSites with insufficient effect size: 513,159\n",
      "\t\tValid sites: 580\n",
      "\t\tSites with significant pvalue (< 0.01): 489\n",
      "\t\tSites with significant FDR adj pvalue (< 0.01): 484\n",
      "(pycoMeth) (pycoMeth) chromosome\tstart\tend\tn_samples\tpvalue\tstatistic\tadj_pvalue\tllr_list\n",
      "1\t119000\t120000\t4\t0.00015113651740556496\t20.243233156522443\t0.0002518941956759416\t-11.65;2.808;-7.75;-7.09\n",
      "1\t515000\t516000\t4\t1.9098737529655853e-09\t43.51853041190212\t9.654105982959306e-09\t2.915;2.965;1.2;-3.815\n",
      "1\t629000\t630000\t4\t9.154292515064807e-07\t30.84714908981499\t2.338982228518761e-06\t4.01;-4.9;-6.71;-5.6\n",
      "1\t1387000\t1388000\t4\t1.838410940117167e-11\t52.99439775910366\t1.5018004862928972e-10\t-4.28;2.21;-2.81;1.63\n",
      "1\t1478000\t1479000\t4\t0.04901293044219966\t7.859204626715323\t0.058253073066548776\t-2.4;-1.34;2.085;-1.62\n",
      "1\t1609000\t1610000\t4\t1.2788076597631697e-16\t77.10996686779535\t9.27135553328298e-15\t2.93;-3.14;0.895;3.51\n",
      "1\t1783000\t1784000\t4\t0.006808685409431452\t12.17450039075581\t0.008894228687996041\t-2.41;1.11;2.34;1.03\n",
      "1\t2611000\t2612000\t4\t9.085991552467002e-11\t49.73764679356497\t6.083248398246381e-10\t-2.605;-4.31;3.035;-2.49\n",
      "1\t3074000\t3075000\t4\t0.00805757437545997\t11.811502137629539\t0.010408447968300183\t2.78;-1.225;2.41;-2.315\n",
      "(pycoMeth) track name=Interval_Medaka_meth_comp itemRgb=On\n",
      "1\t119000\t120000\t.\t3.6\t.\t119000\t120000\t'255,127,0'\n",
      "1\t515000\t516000\t.\t8.02\t.\t515000\t516000\t'255,0,0'\n",
      "1\t629000\t630000\t.\t5.63\t.\t629000\t630000\t'255,0,0'\n",
      "1\t1387000\t1388000\t.\t9.82\t.\t1387000\t1388000\t'255,0,0'\n",
      "1\t1478000\t1479000\t.\t1.23\t.\t1478000\t1479000\t'100,100,100'\n",
      "1\t1609000\t1610000\t.\t14.0\t.\t1609000\t1610000\t'255,0,0'\n",
      "1\t1783000\t1784000\t.\t2.05\t.\t1783000\t1784000\t'255,127,0'\n",
      "1\t2611000\t2612000\t.\t9.22\t.\t2611000\t2612000\t'255,0,0'\n",
      "1\t3074000\t3075000\t.\t1.98\t.\t3074000\t3075000\t'255,127,0'\n",
      "(pycoMeth) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoMeth Meth_Comp \\\n",
    "    -i ./data/Medaka_sample_*_Interval_aggregate.tsv \\\n",
    "    -f ./data/medaka_toplevel.fa \\\n",
    "    -b ./results/CLI_Interval_Medaka.bed \\\n",
    "    -t ./results/CLI_Interval_Medaka.tsv \\\n",
    "    -s Interval_Medaka \\\n",
    "    --progress\n",
    "\n",
    "head ./results/CLI_Interval_Medaka.bed\n",
    "head ./results/CLI_Interval_Medaka.tsv"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "hide_input": false,
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
